src/System.Windows.Forms/src/System/Windows/Forms/TaskDialogIcon.cs

```diff
@@ -93,14 +93,32 @@ public class TaskDialogIcon
 #pragma warning restore IDE1006 // Naming Styles
 
         private readonly TaskDialogStandardIcon? _standardIcon;
-
         private readonly IntPtr? _iconHandle;
 
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="TaskDialogIcon"/> class from an
+        ///   <see cref="Bitmap"/> instance.
+        /// </summary>
+        /// <param name="image">The <see cref="Bitmap"/> instance.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="image"/> is <see langword="null"/>.</exception>
+        /// <remarks>
+        /// <para>
+        ///   The <see cref="Icon"/> instance from which this <see cref="TaskDialogIcon"/>
+        ///   instance is created must not be disposed while the icon is shown in the task
+        ///   dialog.
+        /// </para>
+        /// </remarks>
+        public TaskDialogIcon(Bitmap? image)
+            : this(Icon.FromHandle((image ?? throw new ArgumentNullException(nameof(image))).GetHicon()) ?? default)
```

Comment:
@kpreisser @jeremykuhne @weltkante the docs state we need to [dispose an icon received from `Icon.FromHandle`](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.icon.fromhandle?view=netcore-3.1#remarks).

To achieve this this class will need to be made disposable, and do something like the following:

```cs
    public class TaskDialogIcon : IDisposable
    {
        private readonly Icon? _iconFromBitmap;

        public TaskDialogIcon(Bitmap? image)
            : this(ToIconHandle(image))
		{
		}

		~TaskDialogIcon()
		{
			Dispose(false);
		}

		public void Dispose()
		{
			Dispose(true);
			GC.SuppressFinalizer(this);
		}

		protected void Dispose(bool disposing)
		{
			if (disposing)
				_iconFromBitmap?.Dispose(); // this calls DestroyIcon API.
		}

		private IntPtr ToIconHandle(Bitmap image))
		{
			if (image == null)
				throw new ArgumentNullException(nameof(image));

			_iconFromBitmap = Icon.FromHandle(image.GetHicon());
			return _iconFromBitmap.Handle;
		}
```

Am I overthinking it?

Reply:
I think the docs are wrong. [`Icon.FromHandle()`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Drawing.Common/src/System/Drawing/Icon.Windows.cs#L46) creates an Icon that doesn't dispose of it's `IntPtr`.

Reply:
So, we need to above?

With the tweak:
```cs
		protected void Dispose(bool disposing)
		{
			if (_iconFromBitmap == null)
				return;

			var handle = _iconFromBitmap.Handle;

			if (disposing)
				_iconFromBitmap.Dispose(); 

			DestroyIcon(handle);
		}
```

Reply:
TaskDialogIcon shouldn't have a finalizer if it doesn't have native handles. Since it just has a managed Icon and not a native handle there is no point. If you just wanted to follow the extensibility pattern of Disposable(bool): you forgot to make it virtual so this class is not extensible anyways.

Now regarding the docs for `Icon.FromHandle`. I don't know why @JeremyKuhne thinks the docs are wrong, but `Icon.FromHandle` doesn't take ownership, so whatever handle you put in you need to manually dispose if (and only if) you have ownership. `Image.GetHIcon` is a method which produces a native handle without a disposable wrapper, so you have responsibility to call `DestroyIcon`.

The suggestion from @RussKie may work built its very hard to read and reason about ownership when its spread across multiple things, I generally clone my `Icon.FromHandle` instance to get one that actually maintains ownership, then I can `DestroyIcon` the untracked handle immediately:

```csharp
public static Icon BitmapToIcon(Bitmap bitmap)
{
    IntPtr handle = IntPtr.Zero;
    try
    {
        handle = bitmap.GetHicon();
        var icon = Icon.FromHandle(handle);
        return (Icon)icon.Clone();
    }
    finally
    {
        if (handle != IntPtr.Zero)
            DestroyIcon(handle);
    }
}
```

Reply:
Thank you. I have applies your suggestion.

It still feels as though we rely on GC to clean up after us by not disposing an instance of `TaskDialogIcon`, when a dialog page it is bound to is getting destroyed.
Should we consider making a `TaskDialogPage` dispoable, call `Dispose` from `TaskDialogPage.OnDestroyed`, and dispose `TaskDialogIcon`?

Reply:
Sorry, yes, you should dispose the Icon. I just want to avoid the mess with calling DestroyIcon on a Handle of an Icon because that requires a lot of ownership reasoning. With a cloned Icon you can use normal Dispose. You don't need a Finalizer either since Icon has its own.

Reply:
Ah, thanks for clear it up. I misunderstood you.